src/main/java/bio/terra/filesystem/FireStoreFileDao.java

```diff
@@ -0,0 +1,561 @@
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Paths and document names
+ * FireStore uses forward slash (/) for its path separator. We also use forward slash in our
+ * file system paths. To get uniqueness of objects, we want to name objects with their full
+ * path. Otherwise, two threads could create the same file as two different documents. That
+ * would not do at all.
+ *
+ * We solve this problem by using document names that replace the forward slash in our paths
+ * with character 0x1c - the unicode file separator character. That allows documents to be
+ * named with their full names. (See https://www.compart.com/en/unicode/U+001C)
+ *
+ * Within the document we store the directory path to the object and the object name. That
+ * lets us use the indexes to find the objects in a directory using the index.
+ *
+ * We use FireStore transactions. The required transaction pattern is always read-modify-write.
+ * The transactions are expressed as functions that are retried if another transaction touches
+ * the object between our transaction's read and write.
+ */
+
+
+@Component
+public class FireStoreFileDao {
+    private final Logger logger = LoggerFactory.getLogger("bio.terra.filesystem.FireStoreFileDao");
+
+    private Firestore firestore;
+    private FireStoreUtils fireStoreUtils;
+    private FireStoreDependencyDao dependencyDao;
+
+    @Autowired
+    public FireStoreFileDao(Firestore firestore, FireStoreUtils fireStoreUtils, FireStoreDependencyDao dependencyDao) {
+        this.firestore = firestore;
+        this.fireStoreUtils = fireStoreUtils;
+        this.dependencyDao = dependencyDao;
+    }
+
+    public UUID createFileStart(FSObject fileToCreate) {
+        if (fileToCreate.getObjectType() != FSObject.FSObjectType.INGESTING_FILE) {
+            throw new InvalidFileSystemObjectTypeException("Invalid file system object type");
+        }
+        UUID studyId = fileToCreate.getStudyId();
+        FireStoreObject createObject = makeFileObjectFromFSObject(fileToCreate);
+
+        ApiFuture<UUID> transaction = firestore.runTransaction(xn -> {
+            List<FireStoreObject> createList = new ArrayList<>();
+
+            // Walk up the directory path, finding missing directories we get to an existing one
+            for (String testPath = getDirectoryPath(fileToCreate.getPath());
+                 !testPath.isEmpty();
+                 testPath = getDirectoryPath(testPath)) {
+
+                DocumentSnapshot docSnap = lookupByObjectPath(studyId.toString(), testPath, xn);
+                if (docSnap.exists()) {
+                    break;
+                }
+
+                FireStoreObject dirToCreate = makeDirectoryObject(
+                    studyId,
+                    testPath,
+                    createObject.getFlightId());
+
+                createList.add(dirToCreate);
+            }
+
+            for (FireStoreObject dirToCreate : createList) {
```

Comment:
I went through and marked all of the split points. Also noted that the split point happens in deleteFileWorker to avoid future mistakes in those callers.